### ISSUE TITLE
High number of java.getPackageData calls when opening a java file

### ISSUE THREAD

**Issue by fbricon** [microsoft/vscode-java-dependency#231]
When opening the GreetingResource.java file of a quarkus project, I see java.getPackageData being called 8 times. That seems suboptimal:

```
[Info  - 10:24:43 AM] 4-Feb-2020 10:24:43 AM >> document/didOpen
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> document/codeAction
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> workspace/executeCommand java.maven.initializeSearcher
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> workspace/executeCommand java.resolvePath
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> document/documentSymbol
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> document/foldingRange
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> workspace/executeCommand vscode.java.test.search.codelens
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> workspace/executeCommand vscode.java.resolveMainMethod
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> document/codeAction
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> workspace/executeCommand java.maven.initializeSearcher
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> workspace/executeCommand microprofile/java/codeLens
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> document/foldingRange
[Info  - 10:24:44 AM] 4-Feb-2020 10:24:44 AM >> workspace/executeCommand java.project.list
[Info  - 10:24:46 AM] 4-Feb-2020 10:24:46 AM Reconciled 0. Took 890 ms
[Info  - 10:24:46 AM] 4-Feb-2020 10:24:46 AM Starting com.microsoft.java.debug.plugin
[Info  - 10:24:46 AM] 4-Feb-2020 10:24:46 AM >> workspace/executeCommand java.getPackageData
[Info  - 10:24:46 AM] 4-Feb-2020 10:24:46 AM >> workspace/executeCommand java.getPackageData
[Info  - 10:24:46 AM] 4-Feb-2020 10:24:46 AM >> workspace/executeCommand java.getPackageData
[Info  - 10:24:46 AM] 4-Feb-2020 10:24:46 AM >> workspace/executeCommand java.getPackageData
[Info  - 10:24:46 AM] 4-Feb-2020 10:24:46 AM >> workspace/executeCommand java.getPackageData
[Info  - 10:24:46 AM] 4-Feb-2020 10:24:46 AM >> workspace/executeCommand java.getPackageData
[Info  - 10:24:46 AM] 4-Feb-2020 10:24:46 AM >> workspace/executeCommand java.getPackageData
[Info  - 10:24:46 AM] 4-Feb-2020 10:24:46 AM >> workspace/executeCommand java.getPackageData
[Info  - 10:24:47 AM] 4-Feb-2020 10:24:47 AM begin problem for /GreetingResource.java
[Info  - 10:24:47 AM] 4-Feb-2020 10:24:47 AM 0 problems reported for /GreetingResource.java
[Info  - 10:24:47 AM] 4-Feb-2020 10:24:47 AM Validated 1. Took 472 ms

```

**Comment by fbricon** [comment]
Seems like the dependencies view is making calls even though it's not open

**Comment by fbricon** [comment]
Any plans for fixing this in the short term?

**Comment by jdneo** [comment]
Hi @fbricon, I'll make sure this bug got fixed in the next sprint(start from next week).

**Comment by jdneo** [comment]
Hi @fbricon,

### TL;DR
After some investigation, the 8 times `java.getPackageData` invocations are caused by the feature: synchronize dependency viewer selection with folder explorer. It should not be a bug but by design.

### Root Cause:
- By default, the dependency explorer will automatically sync on event `changeActiveEditor`, which means, changing the active editor will cause the explorer expand some nodes and switch focus to the certain node.
- To expand and focus to a certain nodes, the extension will call `java.getPackageData` several times to get data from different levels. Given a typical project like springboot-petclinic, the 8 times calls are:
  - 1 call from the project level to get all containers
  - 5 calls from different **source** container level to get the package roots.
    - src/main/java
    - src/main/resource
    - src/test/java
    - target/generated-sources/annotations
    - target/generated-test-sources/test-annotations
  - 1 call from the package root level to get all packages
  - 1 call from the package level to get all classes.

The results of the 8 calls will be cached unless the file opened by users belongs to an un-opened node in explorer.

### Follow up
I think the problem here is that the method name `java.getPackageData` is misleading, since some of the invocations are not getting **package** data. A better name could be `java.getNodeData`

**Comment by fbricon** [comment]
But again, one of my main gripes is it the dependencies view is making calls even though it's not open

**Comment by jdneo** [comment]
@fbricon, hmm, yes that's a valid concern. We should add the check for `this._dependencyViewer.visible` here: https://github.com/microsoft/vscode-java-dependency/blob/master/src/views/dependencyExplorer.ts#L25
